.docker/Dockerfile

```diff
@@ -1,36 +1,40 @@
+# syntax=docker/dockerfile:1.7-labs
+# Above must be the first line in Dockerfile.  It enables the "copy --parents" syntax
 ARG ROS_DISTRO=rolling
 FROM ros:$ROS_DISTRO-ros-base AS ci
 
 ENV DEBIAN_FRONTEND=noninteractive
 
-WORKDIR /root/ws_blue
-COPY . src/blue
-
 # Install apt packages needed for CI
 RUN apt-get -q update \
     && apt-get -q -y upgrade \
     && apt-get -q install --no-install-recommends -y \
-    git \
-    sudo \
     clang \
     clang-format-14 \
     clang-tidy \
     clang-tools \
-    python3-pip \
-    python3-dev \
-    lsb-release \
-    wget \
+    git \
     gnupg \
+    lsb-release \
+    python3-dev \
+    python3-pip \
     software-properties-common \
+    sudo \
+    wget \
     && apt-get autoremove -y \
     && apt-get clean -y \
     && rm -rf /var/lib/apt/lists/*
 
-# Install all ROS dependencies for _just_ blue
-# (we have not imported other repos from .repos files)
+WORKDIR /root/ws_blue/src/blue
+COPY --parents */package.xml .
+COPY blue.repos .
+WORKDIR /root/ws_blue
+
+# Install all ROS dependencies for _just_ blue and packages from blue.repos
 RUN apt-get -q update \
     && apt-get -q -y upgrade \
     && rosdep update \
+    && vcs import src < src/blue/blue.repos \
```

Comment:
Same comment as above

Reply:
This is the critical section.  Essentially should CI contain (A) "rosdeps of blue" or (B) "rosdeps of blue + blue's deps"

There are two criteria:

1.  Does (A) exercise some test case in CI which (B) doesn't?   The only case I can think of is if we proactively install apt install a package which is _not correctly declared_ as a `package.xml` dependency in a `blue.repos` dependency.     I think that only happens if we remove a repo from `blue.repos` and don't regenerate the CI image.   That is:

- `blue.repos` includes package A, which requires `libfoo-dev`, so it's installed in the CI image
- Updated `blue.repos` to remove A but add B which requires `libfoo-dev` but erroneously does not require it in `package.xml`
- If we fail to rebuild the CI image, CI will succeed (incorrectly)
- If we rebuild the CI image, `libfoo-dev` will not be installed, and CI will fail (as expected)

2. How does it affect the size of the CI image?

As an example for 'iron':

- Without `blue.repos` CI image is  2.75GB
- With `blue.repos` CI image is  2.72GB

During testing with iron, I noticed this installed `mavros` (and dependencies) from package.   This may require some adjustment in #220

Reply:
This is the critical section.  Essentially should CI contain (A) "rosdeps of blue" or (B) "rosdeps of blue + blue's deps"

Two potential criteria:

1.  Does (A) exercise some test case in CI which (B) doesn't?   The only case I can think of is if we proactively apt install a package which is _not correctly declared_ as a `package.xml` dependency in a `blue.repos` dependency.     I think that only happens if we remove a repo from `blue.repos` and then don't regenerate the CI image.   That is:

- `blue.repos` includes package X, which requires `libfoo-dev`, so it's installed in the CI image
- Updated `blue.repos` to remove X but add Y which requires `libfoo-dev` but erroneously does not require it in `package.xml`
- If we fail to rebuild the CI image, CI will succeed (incorrectly)
- If we rebuild the CI image, `libfoo-dev` will not be installed, and CI will fail (as expected)

2. How does it affect the size of the CI image?

As an example for 'iron':

- Without `blue.repos` CI image is  2.67 GB
- With `blue.repos` CI image is  2.72 GB

During testing with iron, I noticed this installed `mavros` (and dependencies) from package.   This may require some adjustment in #220 for rolling/jazzy.

Reply:
Ah yes, you're completely right about point 1. I've been convinced :smile:  In fact, I think I remember removing the packages in the `blue.repos` file from the respective `package.xml` files because of this.

Reply:
I'm going to add an issue once this is ready and merged mentioning adding the packages from `blue.repos` into the relevant `package.xml` files.

Reply:
Issue documented in #249